### PR TITLE
[Feature][Cache] Sql cache and partition cache #2581

### PR DESCRIPTION
**Pull request by marising**
1. Analyze what mode of cache can be used by query
2. Query cache before executing query in StmtExecutor
3. Two cache mode, sqlcache and partitioncache, are implemented

## Proposed changes

Describe the big picture of your changes here to communicate to the maintainers why we should accept this pull request. If it fixes a bug or resolves a feature request, be sure to link to that issue.

## Types of changes

What types of changes does your code introduce to Doris?
_Put an `x` in the boxes that apply_

- [] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [] Documentation Update (if none of the other choices apply)
- [] Code refactor (Modify the code structure, format the code, etc...)

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [x] I have create an issue on (Fix #ISSUE), and have described the bug/feature there in detail
- [x] Compiling and unit tests pass locally with my changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] If this change need a document change, I have updated the document
- [x] Any dependent changes have been merged

## Further comments

If this is a relatively large or complex change, kick off the discussion at dev@doris.apache.org by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: fe/fe-core/src/main/java/org/apache/doris/qe/cache/PartitionRange.java [hunk 1]
```diff
@@ -0,0 +1,596 @@
...
+            }
+        }
+        return find;
+    }
+
+    /**
+     * Support left or right hit cache, not support middle.
+     * 20200113-2020115, not support 20200114
+     */
+    public Cache.HitRange diskPartitionRange(List<PartitionSingle> rangeList) {
```
Comment (kangkaisen) [inline fe/fe-core/src/main/java/org/apache/doris/qe/cache/PartitionRange.java:350]: `diskPartitionRange `? need a better name.

File: fe/fe-core/src/main/java/org/apache/doris/qe/cache/RowBatchBuilder.java [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+        keyIndex = 0;
+        keyType = Type.INVALID;
+        rowList = Lists.newArrayList();
+        cachePartMap = new HashMap<>();
+        batchSize = 0;
+        rowSize = 0;
+        dataSize = 0;
+    }
+
+    public void partitionIndex(ArrayList<Expr> resultExpr,
```
Comment (kangkaisen) [inline fe/fe-core/src/main/java/org/apache/doris/qe/cache/RowBatchBuilder.java:62]: need a better name.

File: fe/fe-core/src/main/java/org/apache/doris/qe/cache/PartitionRange.java [hunk 3]
```diff
@@ -0,0 +1,596 @@
...
+        }
+
+        rangeList.add(partitionSingleList.get(begin));
+        rangeList.add(partitionSingleList.get(end));
+        LOG.info("the new range for scan be is [{},{}], hit range", rangeList.get(0).getCacheKey().realValue(),
+                rangeList.get(1).getCacheKey().realValue(), hitRange);
+        return hitRange;
+    }
+
+    public List<PartitionSingle> updatePartitionRange() {
```
Comment (kangkaisen) [inline fe/fe-core/src/main/java/org/apache/doris/qe/cache/PartitionRange.java:386]: Add a comment for this method.

File: fe/fe-core/src/main/java/org/apache/doris/qe/cache/CacheAnalyzer.java [hunk 4]
```diff
@@ -0,0 +1,450 @@
...
+            return slot.getColumnName();
+        }
+        return "";
+    }
+
+    /**
+     * Check the selectStmt and tableRefs always group by partition key
+     * 1. At least one group by
+     * 2. group by must contain partition key
+     * 3. AggregateInfo cannot be distinct agg
```
Comment (kangkaisen) [inline fe/fe-core/src/main/java/org/apache/doris/qe/cache/CacheAnalyzer.java:362]: Why AggregateInfo cannot be distinct agg? if distinct cache result is final result, I think which is OK.

File: fe/fe-core/src/main/java/org/apache/doris/qe/cache/CacheAnalyzer.java [hunk 5]
```diff
@@ -0,0 +1,450 @@
...
+                    getPartitionKeyFromSelectStmt((SelectStmt) queryStmt, partColumn, compoundPredicates);
+                }
+            }
+        }
+    }
+
+    /**
+     * Only support case 1
+     * 1.key >= a and key <= b
+     * 2.key = a or key = b
```
Comment (kangkaisen) [inline fe/fe-core/src/main/java/org/apache/doris/qe/cache/CacheAnalyzer.java:312]: Why don't support key = a?  only one BinaryPredicate.

File: fe/fe-core/src/main/java/org/apache/doris/qe/cache/CacheAnalyzer.java [hunk 6]
```diff
@@ -0,0 +1,450 @@
...
+        }
+    }
+
+    public CacheMode getCacheMode() {
+        return cacheMode;
+    }
+
+    public class CacheTable implements Comparable<CacheTable> {
+        public OlapTable olapTable;
+        public long latestId;
```
Comment (kangkaisen) [inline fe/fe-core/src/main/java/org/apache/doris/qe/cache/CacheAnalyzer.java:129]:
```suggestion
        public long latestPartitionId;
```

File: fe/fe-core/src/main/java/org/apache/doris/qe/cache/PartitionCache.java [hunk 7]
```diff
@@ -0,0 +1,215 @@
...
+import org.apache.doris.qe.RowBatch;
+import org.apache.doris.thrift.TUniqueId;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+import java.util.List;
+
+public class PartitionCache extends Cache {
+    private static final Logger LOG = LogManager.getLogger(PartitionCache.class);
+    private SelectStmt nokeyStmt;
```
[the remark this answers is not in the document]
Comment (marising) [reply fe/fe-core/src/main/java/org/apache/doris/qe/cache/PartitionCache.java:42]: After rewriting, there is no partition key select statement

File: fe/fe-core/src/main/java/org/apache/doris/qe/cache/CacheAnalyzer.java [hunk 8]
```diff
@@ -0,0 +1,450 @@
...
+    public CacheAnalyzer(ConnectContext context, StatementBase parsedStmt, List<ScanNode> scanNodes) {
+        this.context = context;
+        this.parsedStmt = parsedStmt;
+        this.scanNodes = scanNodes;
+        checkCacheConfig();
+    }
+
+    private void checkCacheConfig() {
+        if (Config.cache_enable_sql_mode) {
+            if (context.getSessionVariable().isEnableSqlCache()) {
```
[the remark this answers is not in the document]
Comment (marising) [reply fe/fe-core/src/main/java/org/apache/doris/qe/cache/CacheAnalyzer.java:112]: I understand that getsessionvariable() can obtain session variables and global variables. Session variables have higher priority than global variables. I don't know if I understand correctly.

File: fe/fe-core/src/main/java/org/apache/doris/qe/StmtExecutor.java [hunk 9]
```diff
@@ -575,6 +583,78 @@ private void handleSetStmt() {
         context.getState().setOk();
     }
 
+    private void sendChannel(MysqlChannel channel, List<CacheProxy.CacheValue> cacheValues, boolean hitAll)
```
[the remark this answers is not in the document]
Comment (marising) [reply fe/fe-core/src/main/java/org/apache/doris/qe/StmtExecutor.java:586]: This means whether the query partitions are all hit，so isHitAll is better？

File: fe/fe-core/src/main/java/org/apache/doris/qe/cache/PartitionRange.java [hunk 10]
```diff
@@ -0,0 +1,596 @@
...
+                find = true;
+                break;
+            }
+        }
+        return find;
+    }
+
+    /**
+     * Support left or right hit cache, not support middle.
+     * 20200113-2020115, not support 20200114
```
[the remark this answers is not in the document]
Comment (marising) [reply fe/fe-core/src/main/java/org/apache/doris/qe/cache/PartitionRange.java:348]: It's my problem. I explained it in detail in be, but it's simplified here

File: fe/fe-core/src/main/java/org/apache/doris/qe/StmtExecutor.java [hunk 11]
```diff
@@ -575,6 +583,78 @@ private void handleSetStmt() {
...
+            if (batch.isEos()) {
+                break;
+            }
+        }
+        
+        if (cacheResult != null && cacheAnalyzer.getHitRange() == Cache.HitRange.Right) {
+            sendChannel(channel, cacheResult.getValueList(), false);
+        }
+
+        cacheAnalyzer.updateCache();
```
[the remark this answers is not in the document]
Comment (marising) [reply fe/fe-core/src/main/java/org/apache/doris/qe/StmtExecutor.java:652]: The updateCache method determines whether the background Cache needs to be updated
```
    public void updateCache() {
        if (cacheMode == CacheMode.None || cacheMode == CacheMode.NoNeed) {
            return;
        }
        cache.updateCache();
    }
```

File: fe/fe-core/src/main/java/org/apache/doris/qe/cache/PartitionRange.java (hunk 1 above)
Comment (marising) [reply fe/fe-core/src/main/java/org/apache/doris/qe/cache/PartitionRange.java:350]: buildDiskPartitionRange();

File: fe/fe-core/src/main/java/org/apache/doris/qe/cache/CacheAnalyzer.java (hunk 5 above)
Comment (marising) [reply fe/fe-core/src/main/java/org/apache/doris/qe/cache/CacheAnalyzer.java:312]: Other cases are more complicated, I think the first version is simpler

File: fe/fe-core/src/main/java/org/apache/doris/qe/cache/CacheAnalyzer.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (kangkaisen) [reply fe/fe-core/src/main/java/org/apache/doris/qe/cache/CacheAnalyzer.java:112]: Yes. `Session variables have higher priority than global variables.`.

File: fe/fe-core/src/main/java/org/apache/doris/qe/StmtExecutor.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (kangkaisen) [reply fe/fe-core/src/main/java/org/apache/doris/qe/StmtExecutor.java:586]: I think `sendChannel ` shouldn't know query cache hit logic. which only to know how to send and when finish.

File: fe/fe-core/src/main/java/org/apache/doris/qe/cache/PartitionRange.java (hunk 1 above)
Comment (kangkaisen) [reply fe/fe-core/src/main/java/org/apache/doris/qe/cache/PartitionRange.java:350]: OK

File: fe/fe-core/src/main/java/org/apache/doris/qe/StmtExecutor.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (marising) [reply fe/fe-core/src/main/java/org/apache/doris/qe/StmtExecutor.java:586]: ok，i change hitAll to isEos

File: fe/fe-core/src/main/java/org/apache/doris/qe/cache/CacheAnalyzer.java (hunk 4 above)
Comment (marising) [reply fe/fe-core/src/main/java/org/apache/doris/qe/cache/CacheAnalyzer.java:362]: Yes, It's support COUNT(DISTINCT xxx)

**Review by kangkaisen — DISMISSED** [review]
+1

**Review by kangkaisen — APPROVED** [review]
+1

**Pull request merged**